### PR TITLE
Validate API requests

### PR DESCRIPTION
**Pull request by msebragge**
# PR for issue #142

Closing #142

## What is being addressed

Workspace create API request is validated against template.

## How is this addressed

New functionality added to WorkspaceRepository.

Co-authored-by: Tess Ferrandez <tess_ferrandez@hotmail.com>

### CODE REVIEW

File: management_api_app/tests/test_api/test_routes/test_workspaces.py [hunk 1]
```diff
@@ -102,9 +102,10 @@ async def test_workspaces_post_creates_workspace(create_workspace_item_mock, sav
 @patch("service_bus.service_bus.ServiceBus.send_resource_request_message")
 @patch("api.routes.workspaces.WorkspaceRepository.save_workspace")
 @patch("api.routes.workspaces.WorkspaceRepository.create_workspace_item")
-async def test_workspaces_post_calls_db_and_service_bus(create_workspace_item_mock, save_workspace_mock, send_resource_request_message_mock, app: FastAPI, client: AsyncClient):
+@patch("api.routes.workspaces.WorkspaceRepository._validate_workspace_parameters")
+async def test_workspaces_post_calls_db_and_service_bus(_validate_workspace_parameters_mock, create_workspace_item_mock, save_workspace_mock, send_resource_request_message_mock, app: FastAPI, client: AsyncClient):
     workspace_id = "000000d3-82da-4bfc-b6e9-9a7853ef753e"
-
+    _validate_workspace_parameters_mock.return_value = None
```
Comment (sachinkundu) [inline management_api_app/tests/test_api/test_routes/test_workspaces.py:108]: nit: why _? It's not consistent with other mock names

File: management_api_app/tests/test_db/test_repositories/test_workpaces_repository.py [hunk 2]
```diff
@@ -100,3 +110,56 @@ def test_save_workspace_saves_the_items_to_the_database(cosmos_client_mock):
     workspace_repo.save_workspace(workspace)
 
     workspace_repo.container.create_item.assert_called_once_with(body=workspace)
+
+
+@patch('azure.cosmos.CosmosClient')
+def test_validate_workspace_parameters_no_parameters(cosmos_client_mock):
```
Comment (sachinkundu) [inline management_api_app/tests/test_db/test_repositories/test_workpaces_repository.py:116]: what test is this with no assertion??

Comment (msebragge) [reply management_api_app/tests/test_db/test_repositories/test_workpaces_repository.py:116]: Fixed

File: management_api_app/api/routes/workspaces.py [hunk 3]
```diff
@@ -25,6 +26,8 @@ async def retrieve_active_workspaces(workspace_repo: WorkspaceRepository = Depen
 async def create_workspace(workspace_create: WorkspaceInCreate, workspace_repo: WorkspaceRepository = Depends(get_repository(WorkspaceRepository))) -> WorkspaceIdInResponse:
     try:
         workspace = workspace_repo.create_workspace_item(workspace_create)
+    except WorkspaceValidationError as e:
+        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=e.errors)
```
[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/api/routes/workspaces.py]: Changed

File: management_api_app/tests/test_api/test_routes/test_workspaces.py (hunk 1 above)
Comment (msebragge) [reply management_api_app/tests/test_api/test_routes/test_workspaces.py:108]: private vs public

File: management_api_app/db/repositories/workspaces.py [hunk 4]
```diff
@@ -22,10 +23,56 @@ def __init__(self, client: CosmosClient):
...
+        if s == 'int':
+            return 'integer'
+        return 'invalid'
+
+    @staticmethod
+    def _validate_workspace_parameters(template_parameters: List[Parameter], supplied_request_parameters: dict):
+        # verify all required parameters are given
+        errors = {}
+        missing_required = []
+        system_provided = ["acr_name", "porter_driver", "tfstate_container_name", "tfstate_resource_group_name", "tfstate_storage_account_name"]
```
[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/db/repositories/workspaces.py]: Extracted

File: management_api_app/db/repositories/workspaces.py [hunk 5]
```diff
@@ -22,10 +23,56 @@ def __init__(self, client: CosmosClient):
...
+    @staticmethod
+    def _convert_type_name(s: str) -> str:
+        if s == 'str':
+            return 'string'
+        if s == 'int':
+            return 'integer'
+        return 'invalid'
+
+    @staticmethod
+    def _validate_workspace_parameters(template_parameters: List[Parameter], supplied_request_parameters: dict):
```
[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/db/repositories/workspaces.py]: Extracted

File: management_api_app/db/repositories/workspaces.py [hunk 6]
```diff
@@ -22,10 +23,56 @@ def __init__(self, client: CosmosClient):
...
+        for parameter in supplied_request_parameters:
+            if parameter not in template_parameters_by_name:
+                extra_parameters.append(parameter)
+            else:
+                template_parameter = template_parameters_by_name[parameter]
+                supplied_parameter_type = WorkspaceRepository._convert_type_name(type(supplied_request_parameters[parameter]).__name__)
+                if supplied_parameter_type != template_parameter.type:
+                    wrong_type.append({"parameter": parameter, "expected_type": template_parameter.type, "supplied_type": supplied_parameter_type})
+
+        if missing_required:
```
[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/db/repositories/workspaces.py]: Fixed

File: management_api_app/db/repositories/workspaces.py [hunk 7]
```diff
@@ -22,10 +23,56 @@ def __init__(self, client: CosmosClient):
...
+        return 'invalid'
+
+    @staticmethod
+    def _validate_workspace_parameters(template_parameters: List[Parameter], supplied_request_parameters: dict):
+        # verify all required parameters are given
+        errors = {}
+        missing_required = []
+        system_provided = ["acr_name", "porter_driver", "tfstate_container_name", "tfstate_resource_group_name", "tfstate_storage_account_name"]
+        required_parameters = [parameter for parameter in template_parameters if parameter.required and parameter.name not in system_provided]
+        for parameter in required_parameters:
```
[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/db/repositories/workspaces.py:52]: I don't understand what you mean (pardon my limited python skills)

[the remark this answers is not in the document]
Comment (msebragge) [reply management_api_app/db/repositories/workspaces.py]: Fixed

File: management_api_app/tests/test_api/test_routes/test_workspaces.py (hunk 1 above)
Comment (msebragge) [reply management_api_app/tests/test_api/test_routes/test_workspaces.py]: Fixed

**Review by sachinkundu — APPROVED** [review]

**Pull request merged**